pkg/auth/auth.go

```diff
@@ -20,7 +20,7 @@ type Credentials struct {
 
 // CredentialStrategy obtains Credentials from the request.
 type CredentialStrategy interface {
-	GetCredentials(r *http.Request) (*Credentials, error)
+	GetCredentials(w http.ResponseWriter, r *http.Request) (*Credentials, error)
```

Comment:
@labkode this changes the interface. Different credential strategies might have to send different headers. commit d4e73d0cb98a362964d0ca768338f8fd3f95a352 adds the minimal changes. I would like to submit this (and other changes) as a standalone PRs. When will you be able to merge them?

cc @moscicki

Reply:
@butonic for storing credentials in the response there is the token writer strategy: https://github.com/cernbox/reva/blob/master/cmd/revad/svcs/httpsvcs/handlers/auth/tokenwriter/strategy/header/header.go 
Why that one doesn't work for you?

The credentials strategy has the sole purpose of getting credentials from the request (basic auth, tokens, ...) and token strategy has the purpose to write in the response the derivated token from the credentials (cookie, token, your special header, ...)

Reply:
I am throwing the litmus test at reva ocdavsvc. Litmus starts with an unauthenticated DELETE and MKCOL, but the 401 response does not contain a `WWW-Authenticate: Basic realm=whatever` header, so it does not know how to authenticate:

```
Version string: neon 0.29.6: Bundled build, IPv6, Expat, OpenSSL 1.0.2o  27 Mar 2018.
******* Running test 0: init ********
******* Running test 1: begin ********
HTTP session to http://ocdavsvc:9998 begins.
HTTP session to http://ocdavsvc:9998 begins.
ah_create, for WWW-Authenticate
Running pre_send hooks
Sending request headers:
DELETE /remote.php/webdav/litmus/ HTTP/1.1
User-Agent: litmus/0.13 neon/0.29.6
Keep-Alive:
Connection: TE, Keep-Alive
TE: trailers
Host: ocdavsvc:9998
X-Litmus: basic: 1 (begin)

Sending request-line and headers:
Doing DNS lookup on ocdavsvc...
req: Connecting to 172.24.0.3:9998
Request sent; retry is 0.
[status-line] < HTTP/1.1 401 Unauthorized
[hdr] Date: Mon, 11 Mar 2019 12:04:57 GMT
Header Name: [date], Value: [Mon, 11 Mar 2019 12:04:57 GMT]
[hdr] Content-Length: 0
Header Name: [content-length], Value: [0]
[hdr]
End of headers.
Running post_headers hooks
Running post_send hooks
ah_post_send (#0), code is 401 (want 401), WWW-Authenticate is (none)
Request ends, status 401 class 4xx, error line:
401 Unauthorized
Running destroy hooks.
Request ends.
ah_create, for WWW-Authenticate
Running pre_send hooks
Sending request headers:
MKCOL /remote.php/webdav/litmus/ HTTP/1.1
User-Agent: litmus/0.13 neon/0.29.6
Connection: TE
TE: trailers
Host: ocdavsvc:9998
X-Litmus: basic: 1 (begin)

Sending request-line and headers:
Request sent; retry is 1.
[status-line] < HTTP/1.1 401 Unauthorized
[hdr] Date: Mon, 11 Mar 2019 12:04:57 GMT
Header Name: [date], Value: [Mon, 11 Mar 2019 12:04:57 GMT]
[hdr] Content-Length: 0
Header Name: [content-length], Value: [0]
[hdr]
End of headers.
Running post_headers hooks
Running post_send hooks
ah_post_send (#0), code is 401 (want 401), WWW-Authenticate is (none)
Request ends, status 401 class 4xx, error line:
401 Unauthorized
Running destroy hooks.
Request ends.
```

from https://tools.ietf.org/html/rfc7235#section-3.1
> The server generating a 401 response MUST send a WWW-Authenticate header field (Section 4.1) containing at least one challenge applicable to the target resource.

We could add a `GetChallenge()` to the credentials strategy interface instead of passing the writer, but I think this is a lot more flexible.

Reply:
I see,  hen yes, let's go for the w http.ResponseWriter in the signature as you have it,  there is no background compatibility set yet, so "wild wild west".